Implement changelog plugin

panpeter opened the pull request:
This PR adds changelog plugin initial implementation.

Please check `README.md` on what this plugin exactly does and how to use it.

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/ChangelogProcessor.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    fun removePendingEntries(file: File) {
+        file.listFiles().forEach { it.deleteRecursively() }
```
`file.listFiles()` can return null :)

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/EntryValidator.kt:
```diff
@@ -0,0 +1,54 @@
+        data class Invalid(val brokenRules: List<Rule>) : Result()
+    }
+
+    private data class LocalValidator(
+        val isValid: (String) -> Boolean,
+        val rule: Rule
+    )
```
Currently each `Rule` uses the same function `isValid`. I suggest to have class `Rule` as

```kotlin

sealed class Rule {
    abstract fun isValid(entry: String): Boolean

    data class MaxLength(val limit: Int) : Rule() {
        override fun isValid(entry: String) = entry.length <= limit
    }
    object EndsWithDot : Rule() {
        override fun isValid(entry: String) = entry.endsWith(".")
    }
    object DoesNotEndWithDot : Rule() {
        override fun isValid(entry: String) = !entry.endsWith(".")
    }
}
```

and remove `LocalValidator`. 

In case of keeping `LocalValidator`, I'd suggest to pass lambda as the last parameter into constructor to make its usage more flexible with moving lambda outside of constructor. Additionally I'd simplify naming and rename it into `Validator` given that we don't have external validators :)

panpeter replied:
That's a great suggestion! I think I've overcomplicated it 😅 We should simply allow creating any rule in the plugin configuration.

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/GradleApiExt.kt:
```diff
@@ -0,0 +1,13 @@
+import java.io.File
+
+internal fun DirectoryProperty.asPendingChangelogDir(): File? {
+    val dir = asFile.orNull ?: return null
+    if (!dir.exists()) return null
+    if (!dir.isDirectory) throw GradleException("${dir.path} is not a directory.")
+
+    return dir
```
Nit. I think `when` would be more idiomatic here :)

panpeter replied:
Or we could replace 
```
val dir = asFile.orNull ?: return null
if (!dir.exists()) return null
```
with
```
val dir = asFile.orNull?.takeIf { it.exists() } ?: return null
```
😊

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/KotlinClosure.kt:
```diff
@@ -0,0 +1,16 @@
+    fun doCall(it: T): V? = it.function()
+}
+
+internal fun <T> Any.closureOf(action: T.() -> Unit): Closure<Any?> =
+    KotlinClosure(action, this, this)
```
Can't we use `org.gradle.kotlin.dsl.closureOf` instead of that? :)

JinneeJ reviewed (approved):
Works great! 👏 Left some suggestions :)

Here is some general feedback:

1. It would be great to have an error output in format that it can be parsed by IDE with ability to navigate to the file. I didn't test but probably `e: %full_path_to_file%` would work :)
2. Have you thought about making `commitChangelog` task dependent on `checkChangelog` task?

JinneeJ commented:
Also forgot to mention that it would be great to assign group (e.g. `changelog`) for each task :)

panpeter commented:
Thank you for the review and great suggestions 😊

I've addressed all comments and also made it possible to create any rule during the plugin configuration. That's much more simple and flexible: https://github.com/Doist/changelog-gradle-plugin/pull/1/commits/ee34bb21e5e6b15ca34420b9bc76fd57305f36fe

panpeter commented:
> Have you thought about making commitChangelog task dependent on checkChangelog task?

I would leave it for the plugin user to decide when to run the check. For example, if they want to append a pull request link to the pending entries, this could validate some of the rules 😅

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/ChangelogCommitTask.kt:
```diff
@@ -0,0 +1,56 @@
+        val pendingChangelogDir = pendingChangelogDir.asPendingChangelogDir() ?: return
+        val changelogFile = changelogFile.get().asFile
+
+        if (!changelogFile.exists()) {
+            runCatching {
+                changelogFile.mkdirs()
+                changelogFile.createNewFile()
+            }.onFailure {
+                throw GradleException("Cannot create ${changelogFile.name} file.")
+            }
+        }
```
Currently if there is no changelog file, the task fails before this code can run with error:

```> File '%path_to_project%/CHANGELOG.md' specified for property 'changelogFile' does not exist.```

panpeter replied:
Yeah, I think it's just simpler if we remove this part 😅

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/ChangelogProcessor.kt:
```diff
@@ -0,0 +1,34 @@
+        val changelog = changelogFile.readLines().toMutableList()
+        val prefix = config.prefix?.plus("\n") ?: ""
+        val postfix = config.postfix?.plus("\n") ?: ""
+        val entryPrefix = config.entryPrefix ?: ""
+        val entryPostfix = config.entryPostfix ?: ""
```
Probably it's better to add `- ` to `entryPrefix` or `\n` to `entryPostfix`, otherwise by default entries will be rendered as a single paragraph without line breaks.

panpeter replied:
Not sure I understand 🤔

For example, if the file is empty, and we insert `listOf("Feature 1", "Feature 2")` with a default config, the result will be:
```
Feature 1
Feature 2
```
So one entry per line.

I'll add a test to make it easier to play with 👍

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/ChangelogProcessor.kt:
```diff
@@ -0,0 +1,34 @@
+            entryPrefix + it + entryPostfix
+        }
+        changelog.add(config.insertAtLine, entriesBlock)
+        changelogFile.writeText(changelog.joinToString("\n"))
```
Probably we should add an empty line, if original file had it. Otherwise it's removed :)

panpeter replied:
Yes, that's a good point. It's quite tricky that `readLines()` does not include the last empty line.

JinneeJ reviewed (changes requested):
Thanks for updates! Looks good, but there are 2 small blocking issues :)

panpeter commented:
Thank you for another review 😊

I've addressed all comments and pushed two commits.

JinneeJ commented on src/main/kotlin/com/doist/gradle/changelog/ChangelogProcessor.kt:
```diff
@@ -0,0 +1,74 @@
+    fun insertEntries(changelogFile: File, entries: List<String>) {
+        if (entries.isEmpty()) return
+
+        val changelog = LinkedList(changelogFile.readLines())
```
I think it's not optimal to read file several times. Wouldn't it be better to read the file once with `File.readText()` as string, then find line separator in the text and split this string by this line separator. Given that `readText` doesn't skip the last empty line, we won't need additional checks on it :)

panpeter replied:
Yes, that's a much better solution 🤗 Updated.

JinneeJ reviewed (approved):
Thanks for changes! Left one more suggestion regarding reading file 😊

panpeter commented:
Thanks! Update the code and will now merge 👍
